src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,923 @@
+        }
+      }
+    }
+  }
+}
+#endif
+
+VoxelGrid::VoxelGrid(const Mn::Vector3& voxelSize,
+                     const Mn::Vector3i& voxelGridDimensions) {
+  m_voxelSize = voxelSize;
+  m_voxelGridDimensions = voxelGridDimensions;
+  m_offset = Mn::Vector3(0.0, 0.0, 0.0);
+  const int gridSize =
+      voxelGridDimensions[0] * voxelGridDimensions[1] * voxelGridDimensions[2];
+  bool* b_grid = new bool[gridSize];
+  std::shared_ptr<bool> boundary_grid(b_grid);
+  std::memset(boundary_grid.get(), false, gridSize * sizeof(bool));
+  boolGrids_.insert(std::make_pair("Boundary", boundary_grid));
+}
+
+// Creators for extra voxel grids
+
+void VoxelGrid::addBoolGrid(const std::string& gridName) {
+  if (boolGrids_.find(gridName) != boolGrids_.end()) {
+    // grid exists, simply overwrite
+    Mn::Debug() << "Grid" << gridName << "exists, overwriting.";
+    std::memset(boolGrids_[gridName].get(), 0,
+                m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                    m_voxelGridDimensions[2] * sizeof(bool));
+    return;
+  }
+  const int gridSize = m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                       m_voxelGridDimensions[2];
+  bool* n_grid = new bool[gridSize];
+  std::shared_ptr<bool> new_grid(n_grid);
+  std::memset(new_grid.get(), false, gridSize * sizeof(bool));
+  boolGrids_.insert(std::make_pair(gridName, new_grid));
+}
+
+void VoxelGrid::addIntGrid(const std::string& gridName) {
+  if (intGrids_.find(gridName) != intGrids_.end()) {
+    // grid exists, simply overwrite
+    Mn::Debug() << "Grid" << gridName << "exists, overwriting.";
+    std::memset(intGrids_[gridName].get(), 0,
+                m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                    m_voxelGridDimensions[2] * sizeof(int));
+    return;
+  }
+  const int gridSize = m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                       m_voxelGridDimensions[2];
+  int* n_grid = new int[gridSize];
+  std::shared_ptr<int> new_grid(n_grid);
+  std::memset(new_grid.get(), 0, gridSize * sizeof(int));
+  intGrids_.insert(std::make_pair(gridName, new_grid));
+}
+
+void VoxelGrid::addFloatGrid(const std::string& gridName) {
+  if (floatGrids_.find(gridName) != floatGrids_.end()) {
+    // grid exists, simply overwrite
+    Mn::Debug() << "Grid" << gridName << "exists, overwriting.";
+    std::memset(floatGrids_[gridName].get(), 0,
+                m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                    m_voxelGridDimensions[2] * sizeof(float));
+    return;
+  }
+  const int gridSize = m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                       m_voxelGridDimensions[2];
+  float* n_grid = new float[gridSize];
+  std::shared_ptr<float> new_grid(n_grid);
+  std::memset(new_grid.get(), 0, gridSize * sizeof(float));
+  floatGrids_.insert(std::make_pair(gridName, new_grid));
+}
+
+void VoxelGrid::addVector3Grid(const std::string& gridName) {
```

Comment:
a lot of code duplication in addIntGrid, addFloatGrid, addVector3Grid. Can they all share a single template helper function?